pytype_reporter.py

```diff
@@ -499,23 +509,79 @@ def load_config(config_file: str, script_basename: str) -> Config:
     return config
 
 
-def get_changed_xfail_files(config: Config) -> list[str]:
-    """Get the list of changed files compared to the default branch.
-
-    :param config: The configuration dictionary.
-    :return: The list of changed files.
-    """
+def git_diff(*args) -> str:
+    """Run git diff with the given arguments and return the output as a string."""
     return check_output(  # nosec:B603
         args=[
             "git",
             "diff",
             "--ignore-space-change",
-            "--name-only",
-            "origin/" + config["default_branch"],
-            *config["expected_to_fail"],
+            *args,
         ],
         universal_newlines=True,
-    ).splitlines()
+    )
+
+
+def find_branch_point(config):
+    """Get the commit hash of the default where the current branch was created from"""
+    cmd = ["git", "merge-base", "origin/" + config["default_branch"], "HEAD"]
+    return check_output(cmd, universal_newlines=True).strip()  # nosec:B603
+
+
+def github_get_pr_commit_messages(repo, pr_number):
+    """Get the commit messages of the PR from the GitHub API"""
+    empty = []
+    github_token = env.get("GITHUB_TOKEN")
+    debug("GITHUB_TOKEN: %s", github_token)
+    debug("PR_NUMBER: %s", pr_number)
+    if not github_token or not pr_number:
+        return empty
+
+    headers = {
+        "Authorization": f"Bearer {github_token}",
+        "Accept": "application/vnd.github.v3+json",
+    }
+    url = env.get("GITHUB_API_URL") + f"/repos/{repo}/pulls/{pr_number}/commits"
+    debug(url)
+    req = request.Request(url, headers=headers)
+
+    with request.urlopen(req) as response:
+        if response.getcode() == 200:
+            commits = json.loads(response.read())
+            return [commit['commit']['message'].split("\n")[0] for commit in commits]
+        print(f"Failed to fetch PR commits. Status code: {response.getcode()}")
+        print(response.read())
+    return empty
+
+
+def check_only_reverts_from_branch_point(config: Config, changed_files: List[str]):
+    """Check if the branch only contains revert commits since the branch point."""
+
+    repo = env.get("GITHUB_REPOSITORY", "")
+    pr_number = env.get("PR_NUMBER", "")
+    msgs = github_get_pr_commit_messages(repo, pr_number)
+    if not msgs:
+        old_cmd = ["git", "log", "--pretty=%s", find_branch_point(config) + "..HEAD"]
+        msgs = check_output(old_cmd, universal_newlines=True).split("\n")  # nosec:B607 B603
+
+    for commit_message in msgs:  # Check if each commit is a revert
+        print("#> " + commit_message)
+        if commit_message and "Revert" not in commit_message:
+            return False
```

Comment:
Hi @bernhardkaindl , thanks for the fix for the build failure in this PR!
I understand your fix is try to avoid the ci failure for any reverted commit:
1. try to identify the revert commit by search "Revert" in its commit message
2. if it is identified as revert commit, compare it with the code 4 weeks ago

I have 2 thoughts here as this is not temporary code and will be merged into master:
1. Some of the "revert" might be done manually instead of the expected "git revert", just as the first commit in this PR, the commit message might not include "Revert". Can we just skip this judgement and do step 2 directly for all commits? Or will it cause a big ci performance downgrade?
2. For the comparing with code "4 weeks ago", I feel it a little strange as it depends on when the "revert" commit is committed. I think as we will only take the code before the "py2->py3" change as acceptable and not to fail the ci, then can we just compare the code from an exact history point, like the parent of the first commit of the series of py2 -> py3 changes (https://github.com/xapi-project/xen-api/commit/d7f5f3b1e751f35806d44e7faf7292b6c1907926) and its parent is https://github.com/xapi-project/xen-api/commit/ad9f5b13286f04edc3d3e3e7f9c1970fc50d5492
, how about compare just as I did in the [comment](https://github.com/xapi-project/xen-api/pull/5427#issuecomment-1929000084):
git diff <hash in the pr>:scripts/perfmon ad9f5b1:scripts/perfmon